modules/xfem/src/base/XFEM.C

```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
     }
   }
 }
+
+void
+XFEM::saveSolutionForNode(const Node * node_to_store_to,
+                          const Node * node_to_store_from,
+                          SystemBase & sys,
+                          std::map<unique_id_type, std::vector<Real>> & stored_solution,
+                          const NumericVector<Number> & current_solution,
+                          const NumericVector<Number> & old_solution)
+{
+  const std::vector<MooseVariable *> & vars = sys.getVariables(0);
+  _stored_solution_scratch.clear();
+  _stored_solution_scratch.reserve(2 * vars.size());
+  _stored_solution_dofs_scratch.clear();
+  _stored_solution_dofs_scratch.reserve(vars.size());
+  const std::set<SubdomainID> & sids = _moose_mesh->getNodeBlockIds(*node_to_store_from);
+  for (auto var : vars)
+  {
+    if (var->isNodal())
+    {
+      const std::set<SubdomainID> & var_subdomains = sys.getSubdomainsForVar(var->number());
+      std::set<SubdomainID> intersect;
+      set_intersection(var_subdomains.begin(), var_subdomains.end(), sids.begin(), sids.end(),
+                       std::inserter(intersect, intersect.begin()));
+      if (var_subdomains.empty() || !intersect.empty())
+      {
+        unsigned int node_dof = node_to_store_from->dof_number(sys.number(), var->number(), 0);
+        _stored_solution_dofs_scratch.push_back(node_dof);
+      }
+    }
+  }
+
+  for (auto dof : _stored_solution_dofs_scratch)
+    _stored_solution_scratch.push_back(current_solution(dof));
+
+  for (auto dof : _stored_solution_dofs_scratch)
+    _stored_solution_scratch.push_back(old_solution(dof));
```

Comment:
Alternative self documenting data structure below. I'm fine with either but you should at least mention what you are doing.

```
struct DofValues
{
  Real current;
  Real old;
  Real older;
}

std::map<unique_id_type, DofValues> data;
```